File: Day 10/day10.ts
function compile(instructions:string[]): number | null{
    const registros:Record<string, number> = {};
    let i:number = 0;

    while (i < instructions.length){
        const instr = instructions[i].split(' ')
        const f = instr[0];
        const arg = instr.slice(1);

        if (f === 'MOV'){
            const x = arg[0];
            const y = arg[1];
            registros[y] = isNaN(Number(x)) ? registros[x] : Number(x);
        }
        else if (f === 'INC'){
            registros[arg[0]] = (registros[arg[0]] || 0) + 1
        }
        else if (f === 'DEC'){
            registros[arg[0]] = (registros[arg[0]] || 0) - 1
        }
        else if (f === 'JMP'){
            const x = arg[0];
            const y = arg[1];

            if ((registros[x] || 0) === 0){ 
                i = Number(y) - 1;
            }
        }

        i++;
    }

    return 'A' in registros ? registros['A'] : null;
}

File: Day 10/day10.py
def compile(instructions:list[str]) -> int:
  registros = {}
  i = 0
  while i < len(instructions):
    inst = instructions[i].split()
    f, args = inst[0], inst[1:]

    if f == 'MOV':
        x,y = args[0], args[1]
        registros[y] =  registros[x] if x.isalpha() else int(x)

    elif f == 'INC':
        registros[args[0]] = registros.get(args[0],0) + 1
    
    elif f == 'DEC':
        registros[args[0]] = registros.get(args[0],0) - 1
    
    elif f == 'JMP':
        x,y = args[0], args[1]
        if registros.get(x,0) == 0:
            i = int(y) - 1

    i+=1

  return registros.get('A') if 'A' in registros else None